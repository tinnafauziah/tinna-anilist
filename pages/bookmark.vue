<template lang="pug">
.index-anime
  .display-2.mb-4 Bookmarked Anime
  v-layout.scroll_container(wrap, @scroll='handleScroll(debounceLoadAnimeList)')
    template(v-if='isFirstLoading')
      v-card.mb-4.mr-4(v-for="index in 8" :key="index", width='350')
        v-skeleton-loader(
          class='mx-auto',
          height='540',
          width='350',
          type='card'
        )
    template(v-else)
      v-card.mb-4.mr-4.anime_detail_card(v-for='anime in animes', width='350', :key='anime.id', @click='redirectAnimeDetail(anime)')
        v-img(height='400', :src='anime.coverImage.large')
        v-card-text
          .subtitle-1.font-weight-bold.mb-1 {{ anime.title | animeTitle }}
          v-layout(wrap, align-center)
            v-flex.xs1.mr-1
              v-icon(:color='getRatingColor(anime.averageScore)') {{ getRatingIcon(anime.averageScore) }}
            v-flex.xs10
              span.subtitle-1 {{ anime.averageScore | animeAverageScore }}
        v-card-actions
          v-btn(text, color='teal accent-4') Detail
    template(v-if='hasNextPage')
      v-card.mb-4.mr-4(v-for="index in 8" :key="index", width='350')
          v-skeleton-loader(
            class='mx-auto'
            height='540',
            width='350',
            type='card'
          )
</template>

<script>
import { debounce } from "lodash-es";
import humanizeAnimeString from "~/mixins/humanize-anime-string";

export default {
  name: "IndexPage",
  mixins: [humanizeAnimeString],
  data() {
    return {
      isLoadingAnimeList: false,
      isFirstLoading: false,
      isLazyLoading: false,
      hasNextPage: false,
      animes: [],
      page: 1,
    };
  },
  async mounted() {
    this.fetchAnimeBookmarks();
  },
  methods: {
    async fetchAnimeBookmarks(isFirst = true) {
      if (isFirst) {
        this.isFirstLoading = true;
        this.page = 1;
        this.animes = [];
      } else this.isLazyLoading = true;

      let variables = {
        page: this.page,
        perPage: 40,
        sort: this.selectedSort,
      };

      try {
        const response = await this.$store.dispatch(
          "anime/fetchAnimeBookmarks",
          variables
        );

        console.log(response.Viewer.favourites.anime);
        const { nodes, pageInfo } = response.Viewer.favourites.anime;

        this.animes = [...this.animes, ...nodes];
        this.isFirstLoading = true;
        this.hasNextPage = pageInfo.hasNextPage;
        this.page += 1;
      } finally {
        this.isLazyLoading = false;
        this.isFirstLoading = false;
      }
    },
    handleScroll(debounceLoadFunction) {
      const selectedContainer = document.querySelector(".scroll_container");
      if (
        selectedContainer &&
        selectedContainer.scrollHeight <=
          selectedContainer.scrollTop + selectedContainer.offsetHeight + 0.5
      ) {
        debounceLoadFunction();
      }
    },
    debounceLoadAnimeList: debounce(
      function load() {
        this.isLastData = !this.isFirstLoading && !this.hasNextPage;
        if (this.isLazyLoading || !this.hasNextPage) return;
        this.fetchAnimeList(false);
      },
      500,
      { maxWait: 1000 }
    ),
    redirectAnimeDetail(anime) {
      this.$router.push(`/anime/${anime.id}`);
    },
  },
};
</script>
<style lang="scss" scoped>
.scroll_container {
  overflow: scroll;
  height: 586px;
}
</style>
