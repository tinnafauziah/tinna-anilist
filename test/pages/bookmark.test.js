import flushPromises from 'flush-promises';
import bookmark from '~/pages/bookmark';
import animeMock from '~/test/mocks/store-mocks/anime-mocks';

let store;
const mocks = {
  $router: {
    push: jest.fn(),
  },
};
const modules = { ...animeMock };

describe('Bookmark Page', () => {
  const mockPage = () => {
    store = global.createStore({ ...modules });
    return global.shallowMount(bookmark, { store, mocks });
  };
  it('should bookmark page match snapshot', async () => {
    const wrapper = mockPage();
    await flushPromises();
    expect(wrapper.element).toMatchSnapshot();
  });
  it('should redirect to detail', async () => {
    const wrapper = mockPage();

    await flushPromises();

    const selectedDetailCard = await wrapper.find('.anime_detail_card');
    selectedDetailCard.vm.$emit('click');

    expect(mocks.$router.push).toBeCalledWith('/anime/127230');
  });
});
