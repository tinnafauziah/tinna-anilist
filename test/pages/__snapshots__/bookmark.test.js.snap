// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bookmark Page should bookmark page match snapshot, based on current user : { View: [Object] } 1`] = `
<div
  class="index-anime"
>
  <div
    class="display-2 mb-4"
  >
    Bookmarked Anime
  </div>
  <v-layout-stub
    class="scroll_container"
    tag="div"
    wrap="wrap"
  >
    <animecard-stub
      anime="[object Object]"
    />
    <!---->
  </v-layout-stub>
</div>
`;

exports[`Bookmark Page should bookmark page match snapshot, based on current user : null 1`] = `
<div
  class="index-anime"
>
  <div
    class="display-2 mb-4"
  >
    Bookmarked Anime
  </div>
  <v-layout-stub
    class="scroll_container"
    tag="div"
    wrap="wrap"
  >
    <!---->
  </v-layout-stub>
</div>
`;
